src/coinsselectionalgorithm.cpp

```diff
@@ -0,0 +1,361 @@
+{
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    recursions = 0;
+    reachedNodes = 0;
+    reachedLeaves = 0;
+    #endif
+}
+
+CCoinsSelectionBranchAndBound::~CCoinsSelectionBranchAndBound() {
+    delete [] cumulativeAmountsForward;
+}
+
+CAmount* CCoinsSelectionBranchAndBound::PrepareCumulativeAmountsForward()
+{
+    CAmount* cumulativeAmountsForwardTemp = new CAmount[problemDimension + 1];
+    cumulativeAmountsForwardTemp[problemDimension] = 0;
+    for (int index = problemDimension - 1; index >= 0; --index)
+    {
+        cumulativeAmountsForwardTemp[index] = cumulativeAmountsForwardTemp[index + 1] + amounts[index];
+    }
+    return cumulativeAmountsForwardTemp;
+}
+
+void CCoinsSelectionBranchAndBound::SolveRecursive(int currentIndex, size_t tempTotalSize, CAmount tempTotalAmount, unsigned int tempTotalSelection)
+{
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    ++recursions;
+    #endif
+    int nextIndex = currentIndex + 1;
+    // it has been empirically found that it is better to perform first exclusion and then inclusion
+    // this, together with the descending order of coins, is probably due to the fact in this way the algorithm
+    // arrives quickly at exploring tree branches with low amount coins (instead of dealing with included high
+    // amount coins that would hardly represent the optimal solution)
+    for (bool value : { false, true })
+    {
+        if (stopRequested)
+        {
+            break;
+        }
+        tempSelection[currentIndex] = value;
+        #if COINS_SELECTION_ALGORITHM_PROFILING
+        ++reachedNodes;
+        #endif
+        size_t tempTotalSizeNew = tempTotalSize + (value ? sizes[currentIndex] : 0);
+        if (tempTotalSizeNew <= availableTotalSize) // {backtracking}
+        {
+            CAmount tempTotalAmountNew = tempTotalAmount + (value ? amounts[currentIndex] : 0);
+            if (tempTotalAmountNew <= targetAmountPlusOffset) // {backtracking}
+            {
+                CAmount tempTotalAmountNewBiggestPossible = tempTotalAmountNew + cumulativeAmountsForward[nextIndex];
+                if (tempTotalAmountNewBiggestPossible >= targetAmount) // {backtracking}
+                {
+                    unsigned int tempTotalSelectionNew = tempTotalSelection + (value ? 1 : 0);
+                    unsigned int maxTotalSelectionForward = tempTotalSelectionNew + (maxIndex - currentIndex);
+                    if ((maxTotalSelectionForward > optimalTotalSelection) ||
+                        (maxTotalSelectionForward == optimalTotalSelection && tempTotalAmountNewBiggestPossible < optimalTotalAmount)) // {bounding}
+                    {
+                        if (currentIndex < maxIndex)
+                        {
+                            SolveRecursive(nextIndex, tempTotalSizeNew, tempTotalAmountNew, tempTotalSelectionNew);
```

Comment:
Given the number of items in the provided vector of amounts and sizes I would warmly suggest to revrite this function iteratively instead of recursively to avoid the risk to face stack-overflow limits (which are configurable in OS even at small values).

Reply:
I was able to think of this iterative code

```
int selectionSize = 10;
std::vector<int> selection(selectionSize, 0);
int maxIteration = (int)pow(2, selectionSize);
for (int iter = 0; iter < maxIteration; iter++)
    for (int p = 0; p < 32; p++)
        selection[p] = (iter >> p) & 1;
```

but honestly I cannot figure out how to leverage the backtracking and bounding techniques with this iterative way, it would end up listing all the possible combinations with a brute force strategy, something that is not feasible with production data.
By now, what I can do is handle any stackoverflow error with try-catch and possibly reduce the stack frame for increasing recursion depth/capacity (but even right now, considering a stack size of 1 MB, a selection of few more than 16K coins should work fine).

Reply:
Not sure actually if a stackoverflow error can be handled with try-catch; maybe preventing the algorithm running on data sets too big (# > 16K) could be the best approach...

Reply:
You can change the recursion to iteration